IdentityCore/src/webview/operations/MSIDWebResponseBrokerInstallOperation.m

```diff
@@ -0,0 +1,116 @@
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+
+#import "MSIDWebResponseBrokerInstallOperation.h"
+#import "MSIDWebviewResponse.h"
+#import "MSIDBrokerInteractiveController.h"
+#import "MSIDInteractiveTokenRequestParameters.h"
+#import "MSIDWebWPJResponse.h"
+#import "MSIDDefaultTokenRequestProvider.h"
+#import "MSIDKeychainTokenCache.h"
+#import "MSIDLegacyTokenCacheAccessor.h"
+#import "MSIDDefaultTokenCacheAccessor.h"
+#import "MSIDAADV2Oauth2Factory.h"
+#import "MSIDDefaultTokenResponseValidator.h"
+#import "MSIDAccountMetadataCacheAccessor.h"
+
+#if TARGET_OS_IPHONE
+    #import "MSIDAppExtensionUtil.h"
+#endif
+
+@interface MSIDWebResponseBrokerInstallOperation()
+
+@property (nonatomic) NSURL *appInstallLink;
+
+@end
+
+@implementation MSIDWebResponseBrokerInstallOperation
+
+- (nullable instancetype)initWithResponse:(MSIDWebviewResponse *)response
+                                    error:(NSError **)error
+{
+    #if TARGET_OS_IPHONE
+        self = [super initWithResponse:response
+                                 error:error];
+        if (self)
+        {
+            if (![response isKindOfClass:MSIDWebWPJResponse.class] || [NSString msidIsStringNilOrBlank:[(MSIDWebWPJResponse *)response appInstallLink]])
+            {
+                return nil;
+            }
+            
+            MSIDWebWPJResponse *wpjResponse = (MSIDWebWPJResponse *)response;
+            _appInstallLink = [NSURL URLWithString:wpjResponse.appInstallLink];
+        }
+        
+        return self;
+    #else
+        return nil;
+    #endif
+}
+
+- (void)invokeWithInteractiveTokenRequestParameters:(MSIDInteractiveRequestParameters *)interactiveTokenRequestParameters
+                                         completion:(MSIDRequestCompletionBlock)completion
+{
+    if (!completion)
+    {
+        return;
+    }
+    
+    #if TARGET_OS_IPHONE
+        if ([interactiveTokenRequestParameters isKindOfClass:MSIDInteractiveTokenRequestParameters.class])
+        {
+            NSError *brokerError;
+            MSIDBrokerInteractiveController *brokerController = [[MSIDBrokerInteractiveController alloc] initWithInteractiveRequestParameters:(MSIDInteractiveTokenRequestParameters *)interactiveTokenRequestParameters
+                                                                                                                         tokenRequestProvider:[self createTokenRequestProdiving]
```

Comment:
This only allows for a single implementation. Could CPP provide correct token request provider instead?

Reply:
I don't see tokenRequestProvider is provided from CPP side at the moment. I can make the provider as one input parameters and if nothing been passed in we will create ours here otherwise use the CPP version if they are going to provide in the future. Any thought?

Reply:
What I mean is that CPP will need to have that code regardless. Therefore, I'd rather add this code now to CPP, and not add duplicate code here.

Regarding keychain groups - if that is missing right now, it will need to be added to CPP. Let's file a task for tracking.

Reply:
As discussed, I have removed the requestTokenProvider here and will expect to be passed into this operation from CPP side.

Also created this task regarding the keychain groups, feel free to modify/update it:

https://identitydivision.visualstudio.com/Engineering/_workitems/edit/1073184/